plugins/kafka-records-storage/src/main/java/com/github/bsideup/liiklus/kafka/KafkaRecordsStorage.java

```diff
@@ -199,4 +75,142 @@ public int getPartition() {
+                                                return partitionSource;
+                                            })
+                            );
+                        });
+
+                val kafkaReceiver = (DefaultKafkaReceiver<ByteBuffer, ByteBuffer>) KafkaReceiver.create(receiverOptions);
+                receiverRef.set(kafkaReceiver);
+
+                recordsFluxRef.set(
+                        Flux
+                                .defer(kafkaReceiver::receive)
+                                .map(record -> new Record(
+                                        new Envelope(
+                                                topic,
+                                                record.key(),
+                                                record.value()
+                                        ),
+                                        Instant.ofEpochMilli(record.timestamp()),
+                                        record.partition(),
+                                        record.offset()
+                                ))
+                                .share()
+                );
+
+                val disposable = recordsFluxRef.get().subscribe(
+                        null,
+                        assignmentsSink::error,
+                        assignmentsSink::complete
+                );
+
+                assignmentsSink.onDispose(() -> {
+                    disposable.dispose();
+                    DefaultKafkaReceiverAccessor.close(kafkaReceiver);
+                });
+            });
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return this == o;
+        }
+    }
+
+    @Value
+    @ToString(of = "topicPartition")
+    private static class KafkaPartitionSource implements PartitionSource {
+
+        Flux<Record> recordFlux;
+
+        DefaultKafkaReceiver<ByteBuffer, ByteBuffer> kafkaReceiver;
+
+        TopicPartition topicPartition;
+
+        ReplayProcessor<TopicPartition> revocation = ReplayProcessor.create(1);
+
+        AtomicLong requests = new AtomicLong();
+
+        @Override
+        public int getPartition() {
+            return topicPartition.partition();
+        }
+
+        @Override
+        public Publisher<Record> getPublisher() {
+            val partitionList = Arrays.asList(topicPartition);
+            return recordFlux
+                    .filter(it -> it.getPartition() == topicPartition.partition())
+                    .delayUntil(record -> {
+                        if (requests.decrementAndGet() < 0) {
+                            return kafkaReceiver.doOnConsumer(consumer -> {
+                                if (requests.get() < 0) {
+                                    consumer.pause(partitionList);
+                                }
+                                return true;
+                            });
+                        } else {
```

Comment:
last else operator can be omitted - otherwise, it usually adds feeling that something should happen after if - else block, but was forgotten

Reply:
TBH I prefer explicit `if {} else {}` to point out that there are only two branches. `if{} return` is fallthrough-like behaviour and cannot be detected with javac' flow analysis in case you add more branches